src/java/org/apache/cassandra/service/accord/AccordStateCache.java

```diff
@@ -149,17 +147,12 @@ public NamedMap(String name)
         }
     }
 
-    public final Map<Object, Node<?, ?>> active = new HashMap<>();
     private final Map<Object, Node<?, ?>> cache = new HashMap<>();
-    private final Map<Object, WriteOnlyGroup<?, ?>> pendingWriteOnly = new HashMap<>();
-    private final Set<Instance<?, ?>> instances = new HashSet<>();
+    private final Set<Instance<?, ?, ?>> instances = new HashSet<>();
 
-    private final NamedMap<Object, Future<?>> loadFutures = new NamedMap<>("loadFutures");
-    private final NamedMap<Object, Future<?>> saveFutures = new NamedMap<>("saveFutures");
-
-    private final NamedMap<Object, Future<Data>> readFutures = new NamedMap<>("readFutures");
-    private final NamedMap<Object, Future<?>> writeFutures = new NamedMap<>("writeFutures");
+    private final NamedMap<Object, AsyncResult<Void>> saveResults = new NamedMap<>("saveResults");
 
+    private int unreferenced = 0;
     Node<?, ?> head;
```

Comment:
We have a doubly-linked list, and idiomatically these have a dummy header (embedded in the owning class, i.e. so that AccordStateCache would extend Node or some super class). We have templates for this already around the codebase, e.g. IntrusiveLinkedList.

Reply:
Node extends `AccordLoadingState` so it can't also extend some base node class. We also don't want the state cache inheriting the accord loading state stuff.

Reply:
We can invert that hierarchy quite easily can't we? The only class extending it is `Node`, so if we are doing it to separate concerns why not just have a `LoadingNode extends Node` or something?

Reply:
Node also has a bunch of stuff we don’t want on the cache, and I don’t like the idea of AccordLoadingState extending a Node type. It could be done with interfaces, albeit more awkwardly, but I don’t really see the benefit, aside from slightly simpler push/pop methods?

Reply:
Why don't you like inverting the inheritance order?

It's not just that we get simpler push/pop methods: we can reuse the existing intrusive linked lists, so we are duplicating less functionality. It's also more idiomatic, so it's less surprising to a reader familiar with such lists.

Reply:
If you don't want to swap the inheritance order of course, we don't need to have AccordStateCache extend Node, we can just have a dummy member field. It's just very marginally less efficient/neat, and we also avoid the aforementioned code-reuse benefits. 

So, I still think it would be better to inherit since we can. But we can still gain the benefits of a circular-linked list without doing so,

Reply:
> Why don't you like inverting the inheritance order?

Because Node is an implementation detail of the cache, and not the logical ancestor of accord state objects.

> It's also more idiomatic, so it's less surprising to a reader familiar with such lists.

I’d disagree with that. Having explicit head/tail pointers on the list container is _much_ more idiomatic than a container also being a dummy node. You could argue that this does put head/tail fields on the container, but it’s not nearly as straightforward or easy to identify. It does enable cleaner list manipulation methods, but is not the least surprising approach. 

It also breaks the idiomatic linked list traversal pattern, and creates a special node you need to be aware of.

Reply:
🤷 I guess this is a blue/white dress thing. I thought it was pretty widely accepted that sentinel nodes and circular lists were the most common way of implementing doubly-linked lists. Certainly not _singly_ linked lists, in which I would expect separate start/end pointers. But, no point arguing if we don't have the same ground truths there, we will get nowhere.

> Because Node is an implementation detail of the cache, and not the logical ancestor of accord state objects.

I'm confused. The `AccordLoadingState` is _only_ extended by `Node`, and is only used to expose the inner enum `LoadingState` AFAICT outside of the cache hierarchy. What accord state objects would we be the logical ancestor of?

Reply:
> I'm confused. The AccordLoadingState is only extended by Node, and is only used to expose the inner enum LoadingState AFAICT outside of the cache hierarchy. What accord state objects would we be the logical ancestor of?

Well none*, it's a conceptual and structural thing. It keeps things neater and more understandable if they can be kept separate

* a test class does extend it, but it's not really consequential.